### ISSUE TITLE
Understand non-perfect reproducibility in tracks to PV association

### ISSUE THREAD

**Issue by fwyzard** [cms-patatrack/cmssw#397]
During the validation of recent PRs (e.g. #395, #396) we have observed a non-perfect reproducibility of the performance of the pixel tracks associated to the primary vertex.

The results seem to oscillate between these two sets, even when no relevant changes are introduced:

  | development-10824.52 | testing-10824.52
-- | -- | --
Number of TrackingParticles (after cuts) | 4950 | 5017
Number of matched TrackingParticles | 2757 | 2790
Number of tracks | 4371 | 4416
Number of true tracks | 3860 | 3905
Number of fake tracks | 511 | 511
Number of pileup tracks | 0 | 0
Number of duplicate tracks | 0 | 0

The discrepancy is at the 1% level, and the validation plots show only very small differences between the "development" and "testing" points - affecting only a couple of bins, and well within the errors.

**Comment by makortel** [comment]
> Number of TrackingParticles (after cuts) | 4950 | 5017

Is it known why the MC truth gives different numbers?

**Comment by VinInn** [comment]
this is realistic (but running as ideal for that concern the cluster-shape-cut) Quadruplets, isn't' it?
Never seen a discrepancy for "design'?

**Comment by fwyzard** [comment]
Yes, I think I've seen it only in the realistic sample.

**Comment by VinInn** [comment]
could you check the vertex table as in
http://innocent.home.cern.ch/innocent/RelVal/pixOnlyRun3_cpu/plots_vertex.html
I may be enough just one vertex not associated to the MC-truth PV

**Comment by fwyzard** [comment]
Indeed, see [here](https://patatrack.web.cern.ch/patatrack/validation/pulls/7f913d591ce714dae3a9b39161baaab50c6b7d8f/RelValTTbar_13-CMSSW_10_6_0-PU25ns_106X_upgrade2018_realistic_v4-v1/10824.52/plots_vertex.html):

## Pixel vertices

  | development-10824.52 | testing-10824.52
-- | -- | --
Events | 100 | 100
PV reco+tag efficiency | **0.9200** | **0.9300**
Efficiency | 0.5623 | 0.5623
Fake rate | 0.0829 | 0.0829
Merge rate | 0.1144 | 0.1144
Duplicate rate | 0.0069 | 0.0069

## Selected pixel vertices
  | development-10824.52 | testing-10824.52
-- | -- | --
Events | 100 | 100
PV reco+tag efficiency | **0.9500** | **0.9600**
Efficiency | 0.4358 | 0.4358
Fake rate | 0.0249 | 0.0249
Merge rate | 0.1274 | 0.1274
Duplicate rate | 0.0037 | 0.0037


But I do not know if it means one less *associated* vertex, or one less *reconstructed* vertex.

**Comment by fwyzard** [comment]
Same behaviour in the re-validation of #396, and in the validation of #398 .

At this point I wonder if this is simply triggered by re-compiling some extra packages ?

**Comment by fwyzard** [comment]
Mhm, no - `development` and `testing` are checking out the same packages.

**Comment by VinInn** [comment]
efficiency is the same (and other rates as well), so looks association
in principle should be visible in the first of these plots
http://innocent.home.cern.ch/innocent/RelVal/pixOnlyRun3_cpu/plots_pixelVertices/pvtagging.pdf

**Comment by fwyzard** [comment]
Indeed (orange vs black):
![image](https://user-images.githubusercontent.com/4069793/67940922-dbc0e080-fbd4-11e9-8a13-10ddf21b5e17.png)

**Comment by VinInn** [comment]
so in black the real PV is in position 1 instead of 0
it could be sorting (or spitting, even if the number of vertices is the same) or pt of some track...
sum-pt2 is done in parallel in float: it is not stable "by design" (one of the main subject of my lecture on floating point...)
We can accumulate on double...
Still this would mean that the two highest sum-pt2 vertices have very close sum-pt2....
